build-logic/src/main/java/qsl/internal/license/LicenseHeader.java

```diff
@@ -0,0 +1,314 @@
+			this.headerFormat = headerFormat;
+			this.validator = getValidator(headerFormat);
+			this.matcher = getMatcher(headerFormat);
+		}
+
+		public static Rule fromFile(Path path) {
+			byte[] bytes;
+
+			try {
+				bytes = Files.readAllBytes(path);
+			} catch (IOException e) {
+				throw new GradleException(String.format("Failed to load license header %s", path), e);
+			}
+
+			return new Rule(new String(bytes, StandardCharsets.UTF_8));
+		}
+
+		/**
+		 * Returns whether this rule has a special file matching.
+		 *
+		 * @return {@code true} if this rule has a special file matching, otherwise {@code false}
+		 */
+		public final boolean hasSpecialMatching() {
+			return this.matcher != null;
+		}
+
+		/**
+		 * Returns whether the file's licensing should respect this rule.
+		 *
+		 * @param source the source of the file
+		 * @return {@code true} if the file's licensing should respect this rule, otherwise {@code false}
+		 */
+		public boolean match(String source) {
+			if (!this.hasSpecialMatching()) {
+				return true;
+			}
+
+			if (this.matcher.matcher(source).find()) {
+				return true;
+			} else {
+				return this.validator.matcher(source).find();
+			}
+		}
+
+		public boolean validate(String source) {
+			return this.validator.matcher(source).find();
+		}
+
+		public boolean formatFile(Project project, Path rootPath, Path path) {
+			String source = readFile(path);
+			String year = this.getYearString(project, source);
+
+			int delimiter = source.indexOf("package");
+			if (delimiter != -1) {
+				var newSource = this.getLicenseString(year) + source.substring(delimiter);
+
+				if (newSource.equals(source)) {
+					return false;
+				}
+
+				var backupPath = getBackupPath(project, rootPath, path);
+
+				if (backupPath == null) {
+					throw new GradleException("Cannot backup file " + path + ", abandoning formatting.");
+				}
+
+				try {
+					if (!Files.isDirectory(backupPath.getParent())) {
+						Files.createDirectories(backupPath.getParent());
+					}
+
+					Files.copy(path, backupPath, StandardCopyOption.REPLACE_EXISTING);
+				} catch (IOException e) {
+					throw new GradleException("Cannot backup file " + path + ", abandoning formatting.", e);
+				}
+
+				try {
+					Files.writeString(path, newSource);
+				} catch (IOException e) {
+					throw new GradleException("Failed to write updated file " + path + ", abandoning formatting.", e);
+				}
+
+				return true;
+			}
+
+			return false;
+		}
+
+		private String getLicenseString(String year) {
+			var builder = new StringBuilder();
+			var lines = getHeaderLines(this.headerFormat.replace(YEAR_KEY, year).split("\n"));
+
+			builder.append("/*\n");
+			for (var line : lines) {
+				if (line.isEmpty()) {
+					builder.append(" *\n");
+				} else {
+					builder.append(" * ").append(line).append('\n');
+				}
+			}
+
+			builder.append(" */\n\n");
+
+			return builder.toString();
+		}
+
+		private String getYearString(Project project, String source) {
```

Comment:
Can you confirm this matches the openjdk license year behavior, `CREATION YEAR[, LAST MODIFIED]`. I assume this will follow what ever original year exists as well in the file?

Reply:
The minimum year in the file is considered the creation year, so yeah this is possible, though a comma is a bit weird?
You loose all the modifications between last modified and creation year.